File: broadcast-alerts/src/main/java/com/bridgefy/samples/alerts/MainActivity.java
package com.bridgefy.samples.alerts;

import android.Manifest;
import android.bluetooth.BluetoothAdapter;
import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.ActivityCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.TextView;

import com.bridgefy.sdk.client.Bridgefy;
import com.bridgefy.sdk.client.BridgefyClient;
import com.bridgefy.sdk.client.Message;
import com.bridgefy.sdk.client.MessageListener;
import com.bridgefy.sdk.client.RegistrationListener;
import com.bridgefy.sdk.client.StateListener;
import com.bridgefy.sdk.samples.alerts.R;

import java.util.ArrayList;
import java.util.HashMap;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;

public class MainActivity extends AppCompatActivity {

    @BindView(R.id.received_alerts)
    TextView receivedAlerts;
    @BindView(R.id.device_text)
    TextView deviceText;
    @BindView(R.id.device_id)
    TextView deviceId;
    @BindView(R.id.sent_alerts)
    TextView sentAlerts;
    @BindView(R.id.fab)
    FloatingActionButton fab;
    private int sentAlertCounter = 0;
    private int receivedAlertCounter = 0;
    Unbinder unbinder;
    private String TAG = "BRIDGEFY_SAMPLE";
    private ArrayList<Alert> alertsData=new ArrayList<>();
    String fragmentTag = "alerts_fragment";
    String number = "number";
    String date_sent = "date_sent";
    String device_name = "device_name";


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        unbinder = ButterKnife.bind(this);

        deviceText.setText(Build.MANUFACTURER + " " + Build.MODEL);



        BluetoothAdapter bluetoothAdapter = BluetoothAdapter.getDefaultAdapter();


        if (isThingsDevice(this))
        {
            //enabling bluetooth automatically
            bluetoothAdapter.enable();
        }

        Bridgefy.initialize(getApplicationContext(), "68898033-3dce-4c80-843e-e10982b942ac", new RegistrationListener() {
            @Override
            public void onRegistrationSuccessful(BridgefyClient bridgefyClient) {
                super.onRegistrationSuccessful(bridgefyClient);

                //Important data can be fetched from the BridgefyClient object
                deviceId.setText(bridgefyClient.getUserUuid());

                //Once the registration process has been successful, we can start operations
                Bridgefy.start(messageListener, stateListener);
            }

            @Override
            public void onRegistrationFailed(int i, String s) {
                super.onRegistrationFailed(i, s);
            }
        });


        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
    }


    public boolean isThingsDevice(Context context) {
        final PackageManager pm = context.getPackageManager();
        return pm.hasSystemFeature("android.hardware.type.embedded");
    }

    StateListener stateListener = new StateListener() {
        @Override
        public void onStarted() {
            super.onStarted();
            Log.i(TAG, "onStarted: Bridgefy started");


            if (isThingsDevice(getApplicationContext()))
            {
                ///for Android Things devices, we automatically send an alert as soon as possible
                //since there is no UI to do it
                //A message will be dispatched every 10 seconds

                final Handler handler=new Handler();

                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        prepareMessage();
                        handler.postDelayed(this,10000);
                    }
                });


            }
        }

        @Override
        public void onStartError(String s, int i) {
            super.onStartError(s, i);


            switch (i) {

                case (StateListener.INSUFFICIENT_PERMISSIONS):

                    //starting operations will fail if you haven't granted the necessary permissions
                    //request them and try again after they have been granted
                    ActivityCompat.requestPermissions(MainActivity.this,
                            new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, 0);
                    break;
                case (StateListener.LOCATION_SERVICES_DISABLED):
                    //location in the device has been disabled
                    break;


            }
        }

        @Override
        public void onStopped() {
            super.onStopped();
        }
    };


    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {

            //retry again after permissions have been granted
            Bridgefy.start(messageListener, stateListener);
        }
    }



    void prepareMessage()
    {

        sentAlertCounter++;

        //assemble the data that we are about to send


        HashMap<String, Object> data = new HashMap<>();
        data.put(number, sentAlertCounter);
        data.put(date_sent, Double.parseDouble("" + System.currentTimeMillis()));
        data.put(device_name, Build.MANUFACTURER + " " + Build.MODEL);
        Message message = Bridgefy.createMessage(data);


        //Broadcast messages are sent to anyone that can receive it
        Bridgefy.sendBroadcastMessage(message);
        sentAlerts.setText(String.valueOf(sentAlertCounter));
    }

    @OnClick(R.id.fab)
    public void onViewClicked() {


      prepareMessage();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        unbinder.unbind();


        //always stop Bridgefy when it's no longer necessary
        Bridgefy.stop();
    }

    MessageListener messageListener = new MessageListener() {
        @Override
        public void onBroadcastMessageReceived(Message message) {
            super.onBroadcastMessageReceived(message);

            receivedAlertCounter++;
            receivedAlerts.setText(String.valueOf(receivedAlertCounter));



            HashMap<String,Object> content = message.getContent();

            Alert alert = new Alert(message.getSenderId(), (Integer) content.get(number), (String) content.get(device_name), ((Double) content.get(date_sent)).longValue());

            alertsData.add(alert);

            AlertFragment alertFragment= (AlertFragment) getSupportFragmentManager().findFragmentByTag(fragmentTag);
            if (alertFragment!=null)
            {
                alertFragment.updateList(alertsData);
            }
        }

    };


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        if (id == R.id.action_alerts) {

            AlertFragment alertFragment = AlertFragment.newInstance(alertsData);
            getSupportFragmentManager().beginTransaction().add(R.id.content_main,alertFragment,"alerts_fragment").addToBackStack(null).commit();


            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {

        if (getSupportFragmentManager().findFragmentByTag(fragmentTag)!=null)
        {
            getSupportFragmentManager().popBackStackImmediate();
        }
        else {
            super.onBackPressed();
        }
    }
}

File: chat/src/main/java/jp/crinoid/android/MainActivity.kt
package jp.crinoid.android

import android.Manifest
import android.bluetooth.BluetoothAdapter
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.content.LocalBroadcastManager
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.Toolbar
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast

import com.bridgefy.sdk.client.Bridgefy
import com.bridgefy.sdk.client.BridgefyClient
import com.bridgefy.sdk.client.Device
import com.bridgefy.sdk.client.Message
import com.bridgefy.sdk.client.MessageListener
import com.bridgefy.sdk.client.RegistrationListener
import com.bridgefy.sdk.client.Session
import com.bridgefy.sdk.client.StateListener
import jp.crinoid.android.entities.Peer

import java.util.ArrayList
import java.util.HashMap

class MainActivity : AppCompatActivity() {

    private val TAG = "MainActivity"

    internal var peersAdapter = PeersRecyclerViewAdapter(ArrayList())

    private val messageListener = object : MessageListener() {
        override fun onMessageReceived(message: Message?) {
            // direct messages carrying a Device name represent device handshakes
            if (message!!.content["device_name"] != null) {
                val peer = Peer(message.senderId,
                        message.content["device_name"] as String)
                peer.isNearby = true
                peer.deviceType = extractType(message)
                peersAdapter.addPeer(peer)

                // any other direct message should be treated as such
            } else {
                val incomingMessage = message.content["text"] as String
                LocalBroadcastManager.getInstance(baseContext).sendBroadcast(
                        Intent(message.senderId)
                                .putExtra(INTENT_EXTRA_MSG, incomingMessage))
            }

            if (isThingsDevice(this@MainActivity)) {
                //if it's an Android Things device, reply automatically
                val content = HashMap<String, Any>()
                content.put("text", "Beep boop. I'm a bot.")
                val replyMessage = Bridgefy.createMessage(message.senderId, content)
                Bridgefy.sendMessage(replyMessage)

            }
        }

        override fun onBroadcastMessageReceived(message: Message?) {
            // we should not expect to have connected previously to the device that originated
            // the incoming broadcast message, so device information is included in this packet
            val incomingMsg = message!!.content["text"] as String
            val deviceName = message.content["device_name"] as String
            val deviceType = extractType(message)

            LocalBroadcastManager.getInstance(baseContext).sendBroadcast(
                    Intent(BROADCAST_CHAT)
                            .putExtra(INTENT_EXTRA_NAME, deviceName)
                            .putExtra(INTENT_EXTRA_TYPE, deviceType)
                            .putExtra(INTENT_EXTRA_MSG, incomingMsg))
        }
    }

    internal var stateListener: StateListener = object : StateListener() {
        override fun onDeviceConnected(device: Device?, session: Session?) {
            // send our information to the Device
            val map = HashMap<String, Any>()
            map.put("device_name", Build.MANUFACTURER + " " + Build.MODEL)
            map.put("device_type", Peer.DeviceType.ANDROID.ordinal)
            device!!.sendMessage(map)
        }

        override fun onDeviceLost(peer: Device?) {
            Log.w(TAG, "onDeviceLost: " + peer!!.userId)
            peersAdapter.removePeer(peer)
        }

        override fun onStartError(message: String?, errorCode: Int) {
            Log.e(TAG, "onStartError: " + message!!)

            if (errorCode == StateListener.INSUFFICIENT_PERMISSIONS) {
                ActivityCompat.requestPermissions(this@MainActivity,
                        arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), 0)
            }
        }
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // Configure the Toolbar
        val toolbar = findViewById<View>(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)
        toolbar.title = title

        val recyclerView = findViewById<View>(R.id.peer_list) as RecyclerView
        recyclerView.adapter = peersAdapter



        if (isThingsDevice(this)) {
            //enabling bluetooth automatically
            val bluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
            bluetoothAdapter.enable()
        }

        Bridgefy.initialize(applicationContext, object : RegistrationListener() {
            override fun onRegistrationSuccessful(bridgefyClient: BridgefyClient?) {
                // Start Bridgefy
                startBridgefy()
            }

            override fun onRegistrationFailed(errorCode: Int, message: String?) {
                Toast.makeText(baseContext, getString(R.string.registration_error),
                        Toast.LENGTH_LONG).show()
            }
        })
    }

    override fun onDestroy() {
        super.onDestroy()

        if (isFinishing)
            Bridgefy.stop()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_broadcast -> {
                startActivity(Intent(baseContext, ChatActivity::class.java)
                        .putExtra(INTENT_EXTRA_NAME, BROADCAST_CHAT)
                        .putExtra(INTENT_EXTRA_UUID, BROADCAST_CHAT))
                return true
            }
        }
        return false
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
    }


    /**
     * BRIDGEFY METHODS
     */
    private fun startBridgefy() {
        Bridgefy.start(messageListener, stateListener)
    }

    fun isThingsDevice(context: Context): Boolean {
        val pm = context.packageManager
        return pm.hasSystemFeature("android.hardware.type.embedded")
    }

    private fun extractType(message: Message?): Peer.DeviceType {
        val eventOrdinal: Int
        val eventObj = message!!.content["device_type"]
        if (eventObj is Double) {
            eventOrdinal = eventObj.toInt()
        } else {
            eventOrdinal = eventObj as Int
        }
        return Peer.DeviceType.values()[eventOrdinal]
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {

            // Start Bridgefy
            startBridgefy()

        } else if (grantResults[0] == PackageManager.PERMISSION_DENIED) {
            Toast.makeText(this, "Location permissions needed to start peers discovery.", Toast.LENGTH_SHORT).show()
            finish()
        }
    }


    /**
     * RECYCLER VIEW CLASSES
     */
    internal inner class PeersRecyclerViewAdapter(private val peers: MutableList<Peer>) : RecyclerView.Adapter<PeersRecyclerViewAdapter.PeerViewHolder>() {

        override fun getItemCount(): Int {
            return peers.size
        }

        fun addPeer(peer: Peer) {
            val position = getPeerPosition(peer.uuid)
            if (position > -1) {
                peers[position] = peer
                notifyItemChanged(position)
            } else {
                peers.add(peer)
                notifyItemInserted(peers.size - 1)
            }
        }

        fun removePeer(lostPeer: Device?) {
            val position = getPeerPosition(lostPeer!!.userId)
            if (position > -1) {
                val peer = peers[position]
                peer.isNearby = false
                peers[position] = peer
                notifyItemChanged(position)
            }
        }

        private fun getPeerPosition(peerId: String): Int {
            for (i in peers.indices) {
                if (peers[i].uuid == peerId)
                    return i
            }
            return -1
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PeerViewHolder {
            val view = LayoutInflater.from(parent.context)
                    .inflate(R.layout.peer_row, parent, false)
            return PeerViewHolder(view)
        }

        override fun onBindViewHolder(peerHolder: PeerViewHolder, position: Int) {
            peerHolder.settingPeer(peers[position])
        }

        internal inner class PeerViewHolder(view: View) : RecyclerView.ViewHolder(view), View.OnClickListener {

            val mContentView: TextView
            lateinit var peer: Peer

            init {
                mContentView = view.findViewById<View>(R.id.peerName) as TextView
                view.setOnClickListener(this)
            }

            fun settingPeer(peer: Peer) {
                this.peer = peer

                when (peer.deviceType) {
                    Peer.DeviceType.ANDROID -> this.mContentView.text = peer.deviceName + " (android)"

                    Peer.DeviceType.IPHONE -> this.mContentView.text = peer.deviceName + " (iPhone)"
                }

                if (peer.isNearby) {
                    this.mContentView.setTextColor(Color.BLACK)
                } else {
                    this.mContentView.setTextColor(Color.GRAY)
                }
            }

            override fun onClick(v: View) {
                startActivity(Intent(baseContext, ChatActivity::class.java)
                        .putExtra(INTENT_EXTRA_NAME, peer.deviceName)
                        .putExtra(INTENT_EXTRA_UUID, peer.uuid))
            }
        }
    }

    companion object {

        internal val INTENT_EXTRA_NAME = "peerName"
        internal val INTENT_EXTRA_UUID = "peerUuid"
        internal val INTENT_EXTRA_TYPE = "deviceType"
        internal val INTENT_EXTRA_MSG = "message"
        internal val BROADCAST_CHAT = "Broadcast"
    }
}
